pkg/devfile/adapters/common/utils.go

```diff
@@ -20,3 +31,160 @@ func GetSupportedComponents(data versions.DevfileData) []common.DevfileComponent
+// 1. action has to be of type exec 2. component should be present
+func getSupportedCommandActions(command common.DevfileCommand) (supportedCommandActions []common.DevfileCommandAction) {
+	for _, action := range command.Actions {
+		// Check if the command action is of type exec
+		if *action.Type == common.DevfileCommandTypeExec && *action.Component != "" {
+			glog.V(3).Infof("Found command %v for component %v and type %v", command.Name, *action.Component, *action.Type)
+			supportedCommandActions = append(supportedCommandActions, action)
+		}
+	}
+
+	return
+}
+
+// GetRunCommandComponents iterates through the components in the devfile and returns a slice of the corresponding containers
+func GetRunCommandComponents(data versions.DevfileData, devfileRunCmd string) []string {
+	var components []string
+	var emptyCommand common.DevfileCommand
+
+	if devfileRunCmd != "" {
+		command := GetCommand(data, devfileRunCmd)
+		if !reflect.DeepEqual(emptyCommand, command) {
+			for _, action := range command.Actions {
+				components = append(components, *action.Component)
+			}
+		}
+	} else {
+		for _, command := range GetSupportedCommands(data) {
+			if strings.Contains(command.Name, DefaultDevfileRunCommand) {
+				for _, action := range command.Actions {
+					components = append(components, *action.Component)
+				}
+			}
+		}
+	}
+
+	return components
+}
+
+// ValidateAndGetPushDevfileCommands returns a build and run command. It checks if a build
+// or a run command has been explicitly provided during odo push, otherwise it
+// iterates through the devfile commands and returns the default supported command. If neither
+// a build command nor a run command is found, it throws an error
+func ValidateAndGetPushDevfileCommands(data versions.DevfileData, devfileBuildCmd, devfileRunCmd string) ([]common.DevfileCommand, error) {
+	var pushDevfileCommands []common.DevfileCommand
+	var emptyCommand common.DevfileCommand
+	validateBuildCommand := false
+	validateRunCommand := false
+
+	devfileSupportedCommands := GetSupportedCommands(data)
+
+	// Validate the build command if it was provided during odo push
+	if devfileBuildCmd != "" {
+		devfileCommand := GetCommand(data, devfileBuildCmd)
+		if !reflect.DeepEqual(emptyCommand, devfileCommand) {
+			pushDevfileCommands = append(pushDevfileCommands, devfileCommand)
+			validateBuildCommand = true
+		}
+	}
+
+	// Validate the run command if it was provided during odo push
+	if devfileRunCmd != "" {
+		devfileCommand := GetCommand(data, devfileRunCmd)
+		if !reflect.DeepEqual(emptyCommand, devfileCommand) {
+			pushDevfileCommands = append(pushDevfileCommands, devfileCommand)
+			validateRunCommand = true
+		}
+	}
+
+	// If neither is validated, iterate the devfile commands to see for a list of supported commands
+	for _, supportedCommand := range devfileSupportedCommands {
+		if reflect.DeepEqual(supportedCommand.Name, DefaultDevfileBuildCommand) && !validateBuildCommand {
```

Comment:
We can safely compare strings directly in Go, so we shouldn't need the `refelct.DeepEqual` for comparing the string names